engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/types/TypeInferenceSignatures.java

```diff
@@ -0,0 +1,195 @@
+
+import java.util.List;
+import org.enso.compiler.context.InlineContext;
+import org.enso.compiler.context.ModuleContext;
+import org.enso.compiler.core.ir.Expression;
+import org.enso.compiler.core.ir.Function;
+import org.enso.compiler.core.ir.Module;
+import org.enso.compiler.core.ir.Name;
+import org.enso.compiler.core.ir.expression.Application;
+import org.enso.compiler.core.ir.module.scope.Definition;
+import org.enso.compiler.core.ir.module.scope.definition.Method;
+import org.enso.compiler.pass.IRPass;
+import org.enso.compiler.pass.IRProcessingPass;
+import org.enso.compiler.pass.analyse.BindingAnalysis$;
+import org.enso.compiler.pass.resolve.FullyQualifiedNames$;
+import org.enso.compiler.pass.resolve.GlobalNames$;
+import org.enso.compiler.pass.resolve.Patterns$;
+import org.enso.compiler.pass.resolve.TypeNames$;
+import org.enso.compiler.pass.resolve.TypeSignatures$;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.collection.immutable.Seq;
+import scala.collection.immutable.Seq$;
+import scala.jdk.javaapi.CollectionConverters;
+import scala.jdk.javaapi.CollectionConverters$;
+
+// TODO it may make sense to merge this pass into StaticModuleScopeAnalysis, there is little benefit
+// to keeping it separate
+
+/**
+ * A precursor pass that prepares the IR for type inference, run before the main propagation logic
+ * runs in {@link TypeInferencePropagation}.
+ *
+ * <p>It handles storing inferred types based on signatures of the top-level bindings. This is done
+ * as a separate pass, to ensure that once propagation runs, all top-level bindings that have a
+ * signature, already have an inferred type assigned to them, so that these types can be used. This
+ * makes the job of the propagation pass much easier, avoiding it to deal with ensuring in what
+ * order the types are inferred, and ensuring that even without a more complicated unification logic
+ * that will be needed for recursive definitions, many types can already be inferred. In the future,
+ * it may be possible that this pass will no longer be needed - if the propagation pass will be
+ * smart enough to deal with the unknowns. But for now it gives us a very big gain very quickly - as
+ * most of our standard library is annotated with type signatures, we will be able to benefit from
+ * these right away, without needing to implement more complicated recursive inference logic.
+ *
+ * <p>This pass is very simple - it looks at ascribed types of the top level bindings to find out
+ * expected types of each function's arguments, and looks at the outer-most expression of the
+ * binding in search for a return-type ascription (these ascriptions are inserted by the return-type
+ * ascription translation in {@code addTypeAscription} within {@link
+ * org.enso.compiler.core.TreeToIr}). It does not look any deeper into the expressions, ensuring
+ * that it is relatively quick to run.
+ */
+public class TypeInferenceSignatures implements IRPass {
```

Comment:
```suggestion
public final class TypeInferenceSignatures implements IRPass {
```

and also `private` constructor.